### ISSUE TITLE
Recipe-grade common patterns: intent-keyed runnable snippets served through search

### ISSUE THREAD

**Issue by MattGyverLee** [MattGyverLee/FlexToolsMCP#52]
# Spec: Recipe-grade common patterns, served through search

## Motivation

First-draft correctness beats error recovery. Today's `common_patterns_flexicon-v4.1.1.json` (`_schema: common-patterns/1.0`) holds aggregate statistics (`by_object` / `by_operation` counts) — useful for ranking, but it can't hand the model a working snippet. Search (`handle_search_by_capability`, `src/flextoolsmcp/server/handlers/api.py:736`) is keyword+synonym scoring (`DOMAIN_SYNONYMS` api.py:162, `SEARCH_SYNONYMS` api.py:231) with a canonical-intent tier that sorts at score 1000 (api.py:371) — the routing exists; the payload it could route to doesn't.

#45 (invented accessors), #26 (every op is a bare snippet reinvented from scratch), and the skeleton closet (#24) all point the same direction: give the model a verified, runnable starting point for the top ~30 intents.

### 1. Schema: common-patterns/2.0

Add a `recipes` section:

```json
{
  "_schema": "common-patterns/2.0",
  "recipes": {
    "list-entries-with-glosses": {
      "intent": "List all entries with their glosses",
      "match_terms": ["list entries", "dump lexicon", "show glosses"],
      "entities": ["LexEntry", "LexSense"],
      "operations": ["read", "iterate"],
      "requires_write": false,
      "code": "for entry in project.LexEntry.GetAll():\n    headword = project.LexEntry.GetLexemeForm(entry)\n    for sense in project.LexSense.GetAllSenses(entry):\n        gloss = project.LexSense.GetGloss(sense)\n        report.Info(f\"{headword}: {gloss}\")",
      "notes": "GetGloss returns '' for empty (never '***'). GetAll returns EnumerableWrapper; use .Count not len() on the raw collection.",
      "source": "curated | mined",
      "verified_against": {"flexicon": "4.1.2", "verified_by": "eval-corpus"}
    }
  }
}
```

Rules for recipe code: bare-snippet form (the first-class primitive per CLAUDE.md), flexicon imports included where needed, casting already correct, `if modifyAllowed:` guard present when `requires_write`. Every recipe must pass the full preflight chain — enforced by a generation-time check that runs each recipe through the validators (and by inclusion in the eval corpus, tier 1).

### 2. Serving

- `search_by_capability`: when a hit lands in the canonical-intent tier OR a recipe's `match_terms` overlap the expanded query terms, attach the full recipe to that hit as `recipe: {...}` (one recipe max per response, top hit only — recipes are big).
- `find_examples`: `operation_type`/`object_type` filters also search recipes.
- Recipes count as discovery: serving a recipe marks its `entities` via `record_validated_api` — a recipe the server itself verified needs no separate get_object_api round trip.

### 3. Population

- **Curated seed (~20)**: the intents that dominate real sessions — per the operations logs and the FLEXTOOLS-STYLE-GUIDE examples: dump entries/glosses, count senses by POS, find empty definitions, add glosses (write), batch-update citation forms (write), find duplicate headwords, list texts/wordforms, reversal lookups.
- **Mined candidates**: `extract_patterns.py` gains a `--mine-operations-log` mode reading the telemetry JSONL for `outcome: ok` ops with a `user_intent`, clustering by intent, and emitting candidate recipes to a review file. Mined recipes ship only after human review flips `source` to `curated` — never auto-ship user code.
- Versioning follows the existing index convention: `common_patterns_flexicon-v<flexicon_version>.json`, regenerated in the refresh flow, subject to the index-diff gate (docs/STABILIZATION-STRATEGY.md 2a).

### 4. Relationship to skeletons (#24)

Skeletons = session-local, user-specific saved code (`skeleton_storage.py`). Recipes = shipped, curated, version-pinned. Same shape, different trust level; `list_skeletons` and recipe search should present them with the same field names so the model treats both as "known-good starting points."

## Testing / acceptance

- Generation-time: every shipped recipe passes preflight validators (CI check).
- Unit: query matching a recipe's terms returns exactly one `recipe` on the top hit; unrelated queries return none.
- Acceptance metric: for intents covered by recipes, first-pass green rate (telemetry issue) should approach 100%; also expect fewer `invalid_api_chain` rejects.

## Risks

- Stale recipes after flexicon changes: caught by re-running preflight at refresh time + `verified_against` pinning.
- Model pastes write-recipes without thought: `requires_write` recipes always include the modifyAllowed guard and a WARNING note; write gates unchanged.

Related: #24, #26, #45.

**Comment by MattGyverLee** [comment]
## Verification addendum (post-exploration of the API side)

Exploration of the serving path strengthens this spec's motivation and corrects two details:

**1. common_patterns is currently a build artifact with ZERO runtime consumers.** `extract_patterns.py` mines docstring `example` fields from the flexicon API JSON into entries shaped `{class, method, code, description, operation, object_type, source: "docstring"}` under `by_object` / `by_operation` — but no server handler reads the file. The "patterns index" is dead weight today, which makes it the natural home for recipes rather than a new file.

**2. The runtime example infrastructure is `WORKED_EXAMPLES`** (`server/worked_examples.py`): 5 hardcoded examples `{id, title, summary, tags, library, code, see_also}`, served by `search_by_capability` (up to 3 per response, api.py:939) and `find_examples`. Recipes should REPLACE this hardcoded list, not sit beside it: migrate the 5 entries into the `recipes` section of common-patterns/2.0 and make `worked_examples.py` a loader. One curation pipeline, one schema, and the existing `worked_examples` response key / `test_worked_examples.py` coverage carry over.

**3. Canonical-intent tier details:** `CANONICAL_INTENTS` (api.py:206-227) is 8 hardcoded phrase -> (entity, method) mappings matched by exact substring, scored `_CANONICAL_INTENT_SCORE = 1000` and prepended (api.py:913-921). Recipe `match_terms` should be checked at this same tier — and each recipe can declare its canonical (entity, method) so landing a recipe also grows the #45 intent coverage instead of maintaining two phrase lists.

**4. Skeleton relationship (correction):** skeletons are cross-session and auto-captured, not user-saved — `capture_from_code()` runs after every successful execution (execution.py:1041), extracting top-level `def`s into `~/.flextoolsmcp/skeletons.jsonl` with `{name, source, entities, user_intent, captured_at, op_id, session_id, duration_ms}`; `find_examples` already surfaces them (api.py:1041). So the mining pipeline proposed in section 3 largely EXISTS for function-shaped code — a `--mine-skeletons` mode promoting recurring skeletons to recipe candidates is cheaper than mining the operations log from scratch, and the shared `user_intent` field is the clustering key in both cases.
